File: 141-linked-list-cycle/linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head;
        int count = 0;

        while(slow && fast && fast->next) {
            if(slow == fast) {
                if(count) return true;
                else count++;
            }

            slow = slow->next;
            fast = fast->next->next;
        }

        return false;
    }
};

File: 141-linked-list-cycle/linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        HashSet<ListNode> m = new HashSet<>();
        if(head == null) return false;

        ListNode pointer = head;
        while(pointer != null) {
            if(!m.contains(pointer)) {
                m.add(pointer);
            } else {
                return true;
            }

            pointer = pointer.next;
        }

        return false;
    }
}